### PR TITLE
Fix: Visual Studio toolset is passed from settings to the MSBuild helper

### LINKED ISSUE #3766: msbuild not parsing compiler.toolset correctly when building existing solution
To help us debug your issue please explain:
Conan version: 1.8.2
OS is Win10 build 1803 (17134.345)
VS 2017 15.8.1

### Short version

1. `conan create --settings compiler.toolset=X`
doesn't reach the same command line effect as adding toolset= in the command:
2. `msbuild.build('path to sln', toolset=X)`

No. 1 doesn't affect the executed command line, while No. 2 adds
`/m:4 /p:PlatformToolset="X"`

as was expected from No. 1 also.

VS also added to my confusion. it seems the msbuild parameter /p:PlatformToolset doesn't affect existing vcxproj's toolset setting - The PlatformToolset is configured as part of a PropertyGroup for the configuration (like release|x86) and is taken from the vcxproj file instead of being overridden - visual studio also surprised me slightly here in addition..
See this stackoverflow answer for an example:
[https://stackoverflow.com/a/11548055/4182842](https://stackoverflow.com/a/11548055/4182842)

### Long Story (reproduce example)
Take a project and make it fail on toolset v120_xp but succeed with v140_xp.
For example by creating a class with a method:
`char const* what() const noexcept { return "example"; }`

The keyword noexcept is not supported on v120_xp.

now set your .vcxproj file to have v140_xp.

when I run conan create command 
`--settings compiler.toolset=v120_xp`

The project still builds and succeeds - it seems it still runs with v140_xp as the toolset.
I've even tried changing the code from:
`msbuild.build("path to .sln", upgrade_project=False)`

to:
`msbuild.build("path to .sln", upgrade_project=False, toolset="v120_xp")`

to force the msbuild to use this toolset that I want.

My suspicion is the different .vcxproj are not inheriting this value. by forcing toolset="v120_xp" I did manage to see the command line being executed now contained:
`/m:4 /p:PlatformToolset="v120_xp"`

I did find a stackoverflow post in which someone wrote a comment that makes a lot of sense:
[https://stackoverflow.com/a/11548055/4182842](https://stackoverflow.com/a/11548055/4182842)

so 
`/p:Configuration="Release"`
seems to affect the build type (msbuild process even prints it), but PlatformToolset= doesn't...

while
`/p:PlatformToolset="v120_xp"`

does not at all.

- [X] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [X] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

### PR DESCRIPTION
**Pull request by SSE4**
closes #3766 

Changelog: Fix: Visual Studio toolset is passed from settings to the MSBuild helper
Docs: https://github.com/conan-io/docs/pull/XXXX

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

**Comment by SSE4** [comment]
I gonna to update this to use `MSBuild` instead of `build_sln_command` in tests

**Comment by SSE4** [comment]
updated tests to use `MSBuild`

**Comment by lasote** [comment]
Please, solve conflicts.

File: conans/test/unittests/client/build/msbuild_test.py [hunk 1]
```diff
@@ -145,3 +145,52 @@ def custom_properties_test(self):
...
+                           ("v90",),
+                           ("v80",)])
+    def explicit_toolset_test(self, expected_toolset):
+        settings = MockSettings({"build_type": "Debug",
+                                 "compiler": "Visual Studio",
+                                 "compiler.version": "15",
+                                 "arch": "x86_64"})
+        conanfile = MockConanfile(settings)
+        msbuild = MSBuild(conanfile)
+        command = msbuild.get_command("projecshould_flags_testt_file.sln", toolset=expected_toolset)
```
Comment (jgsogo) [inline conans/test/unittests/client/build/msbuild_test.py]: Is there any reason for misspelling it?

File: conans/test/unittests/client/build/msbuild_test.py [hunk 2]
```diff
@@ -145,3 +145,52 @@ def custom_properties_test(self):
...
+                           ("11", "v110_xp")])
+    def custom_toolset_test(self, compiler_version, expected_toolset):
+        settings = MockSettings({"build_type": "Debug",
+                                 "compiler": "Visual Studio",
+                                 "compiler.version": compiler_version,
+                                 "compiler.toolset": expected_toolset,
+                                 "arch": "x86_64"})
+        conanfile = MockConanfile(settings)
+        msbuild = MSBuild(conanfile)
+        command = msbuild.get_command("projecshould_flags_testt_file.sln")
```
Comment (jgsogo) [inline conans/test/unittests/client/build/msbuild_test.py]: idem about spelling

File: conans/client/tools/win.py [hunk 3]
```diff
@@ -93,8 +93,31 @@ def latest_vs_version_installed(output):
...
+                         "8": "v80"}
+
+MSVS_VERSIONS = sorted(list(MSVS_DEFAULT_TOOLSETS.keys()), key=int)
+
+
+def msvs_toolset(settings):
+    toolset = settings.get_safe("compiler.toolset")
+    if not toolset:
+        vs_version = settings.get_safe("compiler.version")
+        try:
```
Comment (danimtb) [inline conans/client/tools/win.py]: this try/except could be simplified with ``toolset = MSVS_DEFAULT_TOOLSETS.get(vs_version)``

File: conans/test/unittests/client/build/msbuild_test.py [hunk 4]
```diff
@@ -145,3 +145,52 @@ def custom_properties_test(self):
                                                                       "MyProp2": "MyValue2"})
         self.assertIn('/p:MyProp1="MyValue1"', command)
         self.assertIn('/p:MyProp2="MyValue2"', command)
+
+    @parameterized.expand([("15", "v141"),
```
Comment (jgsogo) [inline conans/test/unittests/client/build/msbuild_test.py]: Why not `"16", "v141"` too? (same for some other tests)

File: conans/client/generators/visualstudio_multi.py [hunk 5]
```diff
@@ -5,28 +5,18 @@
...
-                               "8": "v80"}
-            try:
-                vs_version = settings.get_safe("compiler.version")
-                toolset = default_toolset[vs_version]
-            except KeyError:
-                raise ConanException("Undefined Visual Studio version %s" % vs_version)
-        self._props.append(("PlatformToolset", toolset))
+        self._props = [("Configuration", settings.get_safe("build_type")),
+                       ("Platform", {'x86': 'Win32', 'x86_64': 'x64'}.get(settings.get_safe("arch"))),
+                       ("PlatformToolset", msvs_toolset(settings))]
```
[the remark this answers is not in the document]
Comment (SSE4) [reply conans/client/generators/visualstudio_multi.py]: changed to use toolset variable

File: conans/client/tools/win.py (hunk 3 above)
Comment (SSE4) [reply conans/client/tools/win.py]: simplified

File: conans/client/tools/win.py [hunk 6]
```diff
@@ -93,8 +93,31 @@ def latest_vs_version_installed(output):
...
+            toolset = MSVS_DEFAULT_TOOLSETS[vs_version]
+        except KeyError:
+            return None
+    return toolset
+
+
 def latest_visual_studio_version_installed(output):
-    for version in reversed(["8", "9", "10", "11", "12", "14", "15"]):
+    for version in reversed(MSVS_VERSIONS):
```
[the remark this answers is not in the document]
Comment (SSE4) [reply conans/client/tools/win.py]: moved

File: conans/test/unittests/client/build/msbuild_test.py (hunk 4 above)
Comment (SSE4) [reply conans/test/unittests/client/build/msbuild_test.py]: added 16

File: conans/test/unittests/client/build/msbuild_test.py (hunk 1 above)
Comment (SSE4) [reply conans/test/unittests/client/build/msbuild_test.py]: fixed spelling, it was copy-paste from another test which had spelling error

File: conans/test/unittests/client/build/msbuild_test.py (hunk 2 above)
Comment (SSE4) [reply conans/test/unittests/client/build/msbuild_test.py]: fixed spelling

File: conans/test/unittests/util/msvs_toolset_test.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# vim: tabstop=8 expandtab shiftwidth=4 softtabstop=4
```
[the remark this answers is not in the document]
Comment (SSE4) [reply conans/test/unittests/util/msvs_toolset_test.py]: removed

**Comment by SSE4** [comment]
@jgsogo default value is still `None`.
however, previous behavior used toolset specified in solution itself, which wasn't correct. 
new behavior uses toolset according to your settings (either settings.toolset, or default toolset for the compiler version).
with old behavior, for instance, we seen projects compiled with Visual Studio 2010 toolset, while settings were set for Visual Studio 2017. it was observed on appveyor, and some users weren't able to compile projects locally, because they never had Visual Studio 2010 installed, so toolset wasn't found on their machine.

File: conans/client/tools/win.py [hunk 8]
```diff
@@ -93,8 +93,27 @@ def latest_vs_version_installed(output):
     return latest_visual_studio_version_installed(output=output)
 
 
+MSVS_DEFAULT_TOOLSETS = {"16": "v141",
```
Comment (memsharded) [inline conans/client/tools/win.py]: I think 16 should be removed until we have better understanding of VS2019, its representation in settings.yml, etc.

**Comment by SSE4** [comment]
@lasote what kind of different place where user specified toolset before?

**Comment by memsharded** [comment]
> @jgsogo default value is still None

Not really. Before, ``/p:PlatformToolset`` was not being added to the command line for most cases, as it was None, unless explicitly defined by user.

This behavior changes, and now, ``/p:PlatformToolset`` is going to be set, almost always, as it will get the default toolset of the compiler.version automatically.

This might be ok, we just need to make sure it is not breaking anyone, and if things like other toolsets (_xp, for example) won't be affected.

**Comment by lasote** [comment]
> @lasote what kind of different place where user specified toolset before?

I don't know, it really doesn't matter, but I would bet there is a way to default a different toolset. Anyway in this case, if the user is not specifying a toolset, I think the right thing to do is not adjusting it. What is the benefit of doing it? You are not following any user input, just guessing a default.

**Comment by SSE4** [comment]
so, if you're not specifying the toolset, and build two projects with some "defaults", you may get one of them built via v110 toolset, and another one with v140_xp toolset, for instance. is it really something we would expect specifying None toolset?

**Comment by SSE4** [comment]
as an example: https://github.com/bincrafters/community/issues/535
currently, compiling with compiler.version=15 and compiler.toolset=None may result in usage of v100 toolset, which is not even available on machine.

**Comment by memsharded** [comment]
> currently, compiling with compiler.version=15 and compiler.toolset=None may result in usage of v100 toolset, which is not even available on machine.

That is really weird. Then the issue is that the SDL2 project contains some toolset for defining v100 hardcoded in its .vcproj files? How is it possible that another toolset is used by default?

You might be totally right, and it totally makes sense to define the toolset always if possible.

**Comment by lasote** [comment]
I need to understand the causes behind that behavior to take a decision.

File: conans/client/tools/win.py (hunk 8 above)
Comment (SSE4) [reply conans/client/tools/win.py]: I've removed all traces of VS2019 (16), as it has its own PR (https://github.com/conan-io/conan/pull/4233)

**Comment by lasote** [comment]
I have been thinking about this and yes, you are right. With the current model, it makes sense to adjust the toolset to the default of the compiler, because in our model the IDE version is defining the toolset to be used (unless other is specified). Please, @SSE4 take a look at the `toolsets_versions` dict declared in the `model/info.py`, I would like to see all the toolset-matching information together.

**Comment by SSE4** [comment]
okay, I'll take a look (that one uses "reversed dict" - keys and values are swapped compare to that we have `conans/client/tools/win.py`, so I am open to suggestion how to do it properly in python having a bi-directional dictionary?)

**Pull request merged**